### PR TITLE
Add GC heap fragmentation counters

### PR DESCRIPTION
**Pull request by sywhang**
Add runtime counters for GC heap fragmentation per generation.

This addresses #31951.

### CODE REVIEW

File: src/coreclr/src/gc/gcpriv.h [hunk 1]
```diff
@@ -2873,6 +2871,8 @@ class gc_heap
     size_t get_current_allocated();
     PER_HEAP_ISOLATED
     size_t get_total_allocated();
+    PER_HEAP_ISOLATED
+    size_t get_total_gen_fragmentation (int gen_number);
```
Comment (Maoni0) [inline src/coreclr/src/gc/gcpriv.h]: >size_t get_total_gen_fragmentation (int gen_number); [](start = 4, length = 52)

I don't understand why this is moved?

Comment (sywhang) [reply src/coreclr/src/gc/gcpriv.h]: It was a protected method.

Comment (Maoni0) [reply src/coreclr/src/gc/gcpriv.h]: there's no access level change between the original place and the new place. plus GCHeap can access protected members of gc_heap. 

---
In reply to: [381707206](https://github.com/dotnet/runtime/pull/32580#discussion_r381707206) [](ancestors = 381707206)

**Comment by Maoni0** [comment]
I'm not sure how one could use this effectively. including @preetamshetty1. for example, LOH/gen2 fragmentation could change wildly right after a gen2 GC is done. would be useful to write some usage examples.

**Pull request closed without merging**